pkg/kubelet/nodeconfig/verify.go

```diff
@@ -0,0 +1,139 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package nodeconfig
+
+import (
+	"crypto/sha256"
+	"fmt"
+	"regexp"
+	"sort"
+
+	apiv1 "k8s.io/kubernetes/pkg/api/v1"
+)
+
+const (
+	sha256Alg = "sha256"
+)
+
+// verifiable is for verifying the integrity of a checkpointed configuration (mainly intended to guard against API object mutation)
+type verifiable interface {
+	verify() (parsable, error)
+}
+
+// verifiableConfigMap is for verifying integrity of config from ConfigMaps
+type verifiableConfigMap struct {
+	cm *apiv1.ConfigMap
+}
+
+// verify checks that re-hashing the data in the ConfigMap results in the hash in the ConfigMap's name.
+// If verification fails, returns an error.
+// If verification succeeds, returns a parsable. This interface can be used to parse the ConfigMap into a KubeletConfiguration.
+func (v *verifiableConfigMap) verify() (parsable, error) {
+	name, alg, expectedHash, err := parseConfigName(v.cm.ObjectMeta.Name)
+	if err != nil {
+		return nil, fmt.Errorf("failed: parse config name %q: %v", err, name)
+	}
+
+	actualHash, err := MapStringStringHash(alg, v.cm.Data)
+	if err != nil {
+		return nil, fmt.Errorf("failed: hash config data: %v", err)
+	}
+
+	if actualHash != expectedHash {
+		return nil, fmt.Errorf("failed: verify config data: alg: %q, expected hash: %q, actual hash: %q", alg, expectedHash, actualHash)
+	}
+
+	return &parsableConfigMap{cm: v.cm}, nil
+}
+
+// capture groups:
+// nameDash: config name substring, with trailing `-`
+// name: config name substring, sans trailing `-`
+// alg: algorithm used to produce the hash value
+// hash: hash value
+const algHashRE = `^(?P<nameDash>(?P<name>[a-z0-9.\-]*){0,1}-){0,1}(?P<alg>[a-z0-9]+)-(?P<hash>[a-f0-9]+)$`
```

Comment:
Parsing this any other way would be better then a regex. This is simpler written with one of:
```
^((?P<name>[a-z0-9.\-]*)-|)(?P<alg>[a-z0-9]+)-(?P<hash>[a-f0-9]+)$
^((?P<name>[a-z0-9.\-]*)-){0,1}(?P<alg>[a-z0-9]+)-(?P<hash>[a-f0-9]+)$
```
Not sure why you nee nameDash. Why do we want to allow configs with empty names? This would be really easy to unit test 😄